### PR TITLE
Added proxy autoconfiguration

### LINKED ISSUE #336: OpenTracing Proxy Auto Configuration
## Detailed Description

Define a module providing a Spring Boot Auto Configuration for `opentracing-proxy`.

## Context

- Integration into Spring Boot ecosystem
- Coupling to Spring is encapsulated in separate module/artifact
- Zalando Conventions over Configuration, i.e. ready to use
- Customizability via Spring Boot configuration and beans

### Module
- `opentracing-proxy` (parent) :heavy_check_mark: 
  - `opentracing-proxy` :heavy_check_mark: 
  - `opentracing-proxy-autoconfigure` :new: 

### Auto Configuration
- `OpenTracingProxyAutoConfiguration`
- `@Primary Tracer proxyTracer(Tracer)`?
- Find registered `Plugin` implementations (inject `@Order`ed list)
- Register pre-defined plugins via config (unless a plugin is registered manually?), e.g.

```yaml
opentracing.proxy:
  log-correlation:
    enabled: true # default?
    baggage:
      - flow_id
  auto-tagging:
    enabled: true # default?
    baggage:
      - flow_id
  tag-propagation:
    enabled: true # default?
    tags:
      - customer_id
      - payment_method
      - sales_channel
      - country
      - language
```

### PR DESCRIPTION
**Pull request by whiskeysierra**
## Motivation and Context

Fixes #336

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

### CODE REVIEW

File: opentracing-proxy/opentracing-proxy-autoconfigure/src/main/java/org/zalando/opentracing/proxy/autoconfigure/ProxyTracerAutoConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import lombok.AllArgsConstructor;
+import org.apiguardian.api.API;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnSingleCandidate;
+import org.springframework.boot.autoconfigure.condition.NoneNestedConditions;
```
Comment (whiskeysierra) [inline opentracing-proxy/opentracing-proxy-autoconfigure/src/main/java/org/zalando/opentracing/proxy/autoconfigure/ProxyTracerAutoConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'org.springframework.boot.autoconfigure.condition.NoneNestedConditions'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701101)

File: opentracing-proxy/opentracing-proxy-autoconfigure/src/main/java/org/zalando/opentracing/proxy/autoconfigure/ProxyTracerAutoConfiguration.java [hunk 2]
```diff
@@ -0,0 +1,132 @@
+package org.zalando.opentracing.proxy.autoconfigure;
+
+import io.opentracing.Tracer;
+import lombok.AllArgsConstructor;
+import org.apiguardian.api.API;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.AutoConfigureAfter;
+import org.springframework.boot.autoconfigure.AutoConfigureBefore;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
```
Comment (whiskeysierra) [inline opentracing-proxy/opentracing-proxy-autoconfigure/src/main/java/org/zalando/opentracing/proxy/autoconfigure/ProxyTracerAutoConfiguration.java]: Codacy Issue found: [Avoid unused imports such as 'org.springframework.boot.autoconfigure.condition.ConditionalOnBean'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701101)

File: opentracing-proxy/opentracing-proxy-autoconfigure/src/test/java/org/zalando/opentracing/proxy/autoconfigure/MultipleTracerTest.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.zalando.opentracing.proxy.ProxyTracer;
+
+import java.util.List;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.everyItem;
+import static org.hamcrest.Matchers.hasSize;
+import static org.hamcrest.Matchers.instanceOf;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.not;
```
Comment (whiskeysierra) [inline opentracing-proxy/opentracing-proxy-autoconfigure/src/test/java/org/zalando/opentracing/proxy/autoconfigure/MultipleTracerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.hamcrest.Matchers.not'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701101)

File: opentracing-proxy/opentracing-proxy-autoconfigure/src/test/java/org/zalando/opentracing/proxy/autoconfigure/MultipleTracerTest.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.context.annotation.Primary;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+import org.zalando.opentracing.proxy.ProxyTracer;
```
Comment (whiskeysierra) [inline opentracing-proxy/opentracing-proxy-autoconfigure/src/test/java/org/zalando/opentracing/proxy/autoconfigure/MultipleTracerTest.java]: Codacy Issue found: [Avoid unused imports such as 'org.zalando.opentracing.proxy.ProxyTracer'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701101)

File: opentracing-proxy/opentracing-proxy/src/main/java/org/zalando/opentracing/proxy/ProxySpan.java [hunk 5]
```diff
@@ -1,17 +1,22 @@
...
 import io.opentracing.Span;
-import io.opentracing.tag.BooleanTag;
-import io.opentracing.tag.StringTag;
+import io.opentracing.Tracer;
 import io.opentracing.tag.Tag;
 import lombok.AllArgsConstructor;
+import lombok.EqualsAndHashCode;
 
 import java.util.Map;
+import java.util.Objects;
```
Comment (whiskeysierra) [inline opentracing-proxy/opentracing-proxy/src/main/java/org/zalando/opentracing/proxy/ProxySpan.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.Objects'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701101)

Comment (whiskeysierra) [inline opentracing-proxy/opentracing-proxy/src/main/java/org/zalando/opentracing/proxy/ProxySpan.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.Objects'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701101)

File: opentracing-proxy/opentracing-proxy/src/main/java/org/zalando/opentracing/proxy/LogCorrelation.java [hunk 6]
```diff
@@ -2,15 +2,24 @@
...
 import org.apiguardian.api.API;
 import org.organicdesign.fp.collections.ImMap;
 import org.organicdesign.fp.collections.PersistentHashMap;
 import org.slf4j.MDC;
 
 import javax.annotation.Nullable;
+import java.util.Collection;
+import java.util.Map;
 import java.util.Optional;
+import java.util.Set;
```
Comment (whiskeysierra) [inline opentracing-proxy/opentracing-proxy/src/main/java/org/zalando/opentracing/proxy/LogCorrelation.java]: Codacy Issue found: [Avoid unused imports such as 'java.util.Set'](https://app.codacy.com/app/whiskeysierra/opentracing-toolbox/pullRequest?prid=4701101)

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Comment by whiskeysierra** [comment]
Had to rebase. Can you take another look? @lukasniemeier-zalando

**Comment by lukasniemeier-zalando** [comment]
:+1:

**Pull request merged**